From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jacques Heunis <jheunis@bloomberg.net>
Date: Mon, 28 Apr 2025 11:43:51 +0100
Subject: [PATCH] Initial commit of Weir QoS

---
 Makefile   |  1 +
 src/hlua.c | 31 +++++++++++++++++++++++++++++++
 2 files changed, 32 insertions(+)

diff --git a/Makefile b/Makefile
index 1959ac479..3a099edbc 100644
--- a/Makefile
+++ b/Makefile
@@ -993,6 +993,7 @@ OBJS += src/mux_h2.o src/mux_h1.o src/mux_fcgi.o src/log.o		\
         src/http_acl.o src/dict.o src/dgram.o src/pipe.o		\
         src/hpack-huff.o src/hpack-enc.o src/ebtree.o src/hash.o	\
         src/version.o src/ncbmbuf.o
+OBJS += src/flt_weir.o
 
 ifneq ($(TRACE),)
   OBJS += src/calltrace.o
diff --git a/src/hlua.c b/src/hlua.c
index 9699936e8..875f6113d 100644
--- a/src/hlua.c
+++ b/src/hlua.c
@@ -14038,6 +14038,35 @@ static void *hlua_alloc(void *ud, void *ptr, size_t osize, size_t nsize)
 	return ptr;
 }
 
+/* Handle receipt of a new limit-share update, instructing the local instance on what limit to apply to a given key.
+ * In lua code, this is called as ingest_weir_limit_share(timestamp_seconds, key, instance, "up"|"down", limit)
+ * Example: ingest_weir_limit_share("1716197521", "myaccesskey", "1f2e3d4c", "down", "4096")
+ */
+int weir_ingest_limit_share_update(uint64_t, const char*, const char*, const char*, uint64_t);
+__LJMP static int hlua_ingest_weir_limit_share_update(lua_State *L)
+{
+	const char* timestamp_str;
+	const char *user_key;
+	const char *instance_id;
+	const char *direction;
+	const char* limit_str;
+	uint64_t timestamp;
+	uint64_t limit;
+	int success;
+
+	timestamp_str = MAY_LJMP(luaL_checkstring(L, 1));
+	user_key = MAY_LJMP(luaL_checkstring(L, 2));
+	instance_id = MAY_LJMP(luaL_checkstring(L, 3));
+	direction = MAY_LJMP(luaL_checkstring(L, 4));
+	limit_str = MAY_LJMP(luaL_checkstring(L, 5));
+	timestamp = strtoull(timestamp_str, NULL, 10);
+	limit = strtoull(limit_str, NULL, 10);
+
+	success = weir_ingest_limit_share_update(timestamp, user_key, instance_id, direction, limit);
+	lua_pushboolean(L, success);
+	return 1;
+}
+
 /* This function can fail with an abort() due to a Lua critical error.
  * We are in the initialisation process of HAProxy, this abort() is
  * tolerated.
@@ -14160,6 +14189,7 @@ lua_State *hlua_init_state(int thread_num)
 	hlua_class_function(L, "Alert", hlua_log_alert);
 	hlua_class_function(L, "done", hlua_done);
 	hlua_class_function(L, "disable_legacy_mailers", hlua_disable_legacy_mailers);
+	hlua_class_function(L, "ingest_weir_limit_share_update", hlua_ingest_weir_limit_share_update);
 	hlua_fcn_reg_core_fcn(L);
 
 	lua_setglobal(L, "core");
-- 
2.43.0

